quantum/process_keycode/process_autocorrect.c

```diff
@@ -48,9 +48,10 @@ bool process_autocorrect(uint16_t keycode, keyrecord_t* record) {
             return true;
 #ifndef NO_ACTION_TAPPING
         case QK_MOD_TAP ... QK_MOD_TAP_MAX:
-            if (((keycode >> 8) & 0xF) == MOD_LSFT) {
+            if (((keycode >> 8) & 0xF) == MOD_LSFT && !record->tap.count) {
                 return true;
             }
+            // fall through intentionally
 #    ifndef NO_ACTION_LAYER
         case QK_LAYER_TAP ... QK_LAYER_TAP_MAX:
 #    endif
```

Comment:
I found a bug in the tap-hold related logic. This conditional

```
if (record->event.pressed || !record->tap.count) {
    return true;
}
```

should instead be

```
if (!record->event.pressed || !record->tap.count) {
    return true;
}
```

That is, we want early return on release rather than on press. If I make my 'T' key a mod-tap, the current logic misbehaves on typos ending in 'T', for instance "ouput" gets improperly autocorrected to "ouptput" instead of "output". Changing the condition to `!record->event.pressed` fixes this.

I believe we can do an early return on any kind of key release. To simplify the logic, I suggest we check for and return on release above the switch. It also seems easier to uniformly test for non shift mods above the switch rather than per each case. In total, here is what I am suggesting:

```
// Ignore key release; we only process key presses.
if (!record->event.pressed) {
    return true;
}
// Disable autocorrect while a mod other than shift is active.
if (((get_mods() | get_oneshot_mods()) & ~MOD_MASK_SHIFT) != 0) {
    typo_buffer_size = 0;
    return true;
}

switch (keycode) {
    case KC_LSFT:
    case KC_RSFT:
        return true;  // Ignore shift keys.
#ifndef NO_ACTION_TAPPING
    case QK_MOD_TAP ... QK_MOD_TAP_MAX:
#    ifndef NO_ACTION_LAYER
    case QK_LAYER_TAP ... QK_LAYER_TAP_MAX:
#    endif
        if (!record->tap.count) {
            return true; // Early return if tap-hold key is not being tapped.
        }
        keycode &= 0xFF;
        break;
#endif
#ifdef SWAP_HANDS_ENABLE
    case QK_SWAP_HANDS ... QK_SWAP_HANDS_MAX:
        // process SH_T(kc) keycodes (0x5600-0x56F0, 0x56F0-0x56FF are special oneshot functions)
        if (keycode > 0x56F0 || !record->tap.count) {
            return true;
        }
        keycode &= 0xFF;
        break;
#endif
#ifndef NO_ACTION_ONESHOT
    case QK_ONE_SHOT_MOD ... QK_ONE_SHOT_MOD_MAX:
        return true;  // Ignore one-shot shift key.
#endif
}
```

Reply:
The problem with the early return is that it will break stuff like layer and mod tap, which you can't determine if it's a hold action or a tap action until later in the processing.  

Grabbing it on release means that you can check the count.  If it has a count it was a tap event, and you can grab the keycode from it.  Otherwise, you may be grabbing a layer/mod action, and erroneously processing it as a tap and grabbing the keycode. 

That said, I do use both layer taps and mod taps in my keymap, and it does seem to work fine, as is.

Reply:
> if (record->event.pressed || !record->tap.count) {

For layer tap and mod tap, this returns true on the press (since we cannot know the  tap-hold state at this point, and we return on 0 for tap.count, as that means that this is a hold, not a tap.

Reply:
Though, I may take a look at trying to write some tests for this (something I've not done before).   That way, we can be sure that all of the functions work correctly, from the software side.

Reply:
Aha, thanks! That's a great point about tap-hold keys not being settled yet on press, I overlooked this. I'm wondering now if tap-hold configurations could also possibly matter in how exactly this logic works, so FWIW, I've been testing this with PERMISSIVE_HOLD + IGNORE_MOD_TAP_INTERRUPT + TAPPING_FORCE_HOLD_PER_KEY.

It sounds like you're saying the way to go is have autocorrect handle each key on *release*, rather than press. Or maybe handle just the tap-hold keys on release, and all other keys on press? This is possible, but an edge case adjustment is needed when making a typo correction: if the typo is detected on a key press (i.e. the typo ends on a regular non tap-hold key), then the last key is "intercepted" before it is sent to the host, so we don't need to backspace that key. But if ending on a key release (i.e. the typo ends on a tap-hold key), the key is already sent to the host, so we need to send one more backspace before the correction.

Specifically, line 148 

```
const uint8_t backspaces = code & 63;
```

should be changed to

```
const uint8_t backspaces = (code & 63) + !record->event.pressed;
```

Reply:
Welcome! 

And yeah, I think that on release for tap-hold keys, and on press for the rest. 

And ... the tap-hold stuff is very convoluted, and a PITA to deal with. So I can absolutely understand getting tripped up on them. 


And applied the suggestion.

Reply:
I did some more testing with mod-taps and found another edge case :-/. The current logic works as intended on clean, non-overlapping taps, but fails to detect typos when "rolling" from a mod-tap key to a regular key.

For example, my 'N' key is a mod-tap and 'G' is a regular key, and let's consider the typo correction entry `cieling -> ceiling`. When typing "`cieling`" normally, the typo is not detected. This is because I will roll the "`ng`" like `n down, g down, n up, g up`. The current logic handles mod-tap keys on release, and normal keys on press, so the 'G' key appends to the buffer before the 'N'. So autocorrect thinks I typed "`cielign`" and doesn't recognize the typo. But the typo is detected if I slow down and type in non-overlapping taps `n down, n up, g down, g up`.

It seems the simplest way to fix this problem is to uniformly handle all keys on release, by changing line 84 in the default case

```
if (!record->event.pressed) {
```

to

```
if (record->event.pressed) {
```

With this change, I can type "`cieling`" quickly and it is properly detected.

Reply:
Have you tried checking `record->tap.interrupted`.  IIRC, it's what happens when you roll keys.

Reply:
Good news! I believe I have a simple fix to handling mod-tap keys. Following your suggestion to look at `tap.interrupted`, I spent some time studying `action_tapping_process()`. I'd looked at it a few times before, it has a mesmerizing number of logical cases. For a bigger picture, I studied how the action_tapping code and process_autocorrect() relate to each other in the call graph.

I realized the life of handling a key event is basically this:

1. [`void action_exec(keyevent_t event)`](https://github.com/qmk/qmk_firmware/blob/d9393b86842b7ef143259b5f771ae7969f98cbb4/quantum/action.c#L70) is the starting point.
2. action_exec() in turn calls into the action_tapping code [void action_tapping_process(keyrecord_t record)](https://github.com/qmk/qmk_firmware/blob/d9393b86842b7ef143259b5f771ae7969f98cbb4/quantum/action_tapping.c#L69). The code here is complicated, and depending on the situation, the key event might go into a waiting buffer before it is handled further.
3. Once the tap-hold key has "settled" the event, it calls [void process_record(keyrecord_t *record)](https://github.com/qmk/qmk_firmware/blob/d9393b86842b7ef143259b5f771ae7969f98cbb4/quantum/action.c#L203). If it is mod-tap key press or release, it appears that the event's `tap.count` field is always set by this point.
4. process_record() in turn calls process_record_quantum(), which is the main point for calling into other features including autocorrect.

Importantly, the only way that process_record() is called is through the action_tapping code, and it will delay making that call if needed to determine whether a mod-tap key is being tapped vs. held. That simplifies things---hurray! I believe that  after all use the strategy I described earlier:
https://github.com/qmk/qmk_firmware/pull/15699#discussion_r785495927

That is, handle all events on key press, and for tap-hold keys, make an early return if `tap.count` is zero:

```
#ifndef NO_ACTION_TAPPING
    case QK_MOD_TAP ... QK_MOD_TAP_MAX:
#    ifndef NO_ACTION_LAYER
    case QK_LAYER_TAP ... QK_LAYER_TAP_MAX:
#    endif
        if (!record->tap.count) {
            return true; // Early return if tap-hold key is not being tapped.
        }
        keycode &= 0xFF;
        break
```